Userland/Libraries/LibJS/Runtime/ObjectPrototype.cpp

```diff
@@ -73,39 +75,71 @@ JS_DEFINE_NATIVE_FUNCTION(ObjectPrototype::to_string)
 {
     auto this_value = vm.this_value(global_object);
 
+    // 1. If the this value is undefined, return "[object Undefined]".
     if (this_value.is_undefined())
         return js_string(vm, "[object Undefined]");
+
+    // 2. If the this value is null, return "[object Null]".
     if (this_value.is_null())
         return js_string(vm, "[object Null]");
 
-    auto* this_object = this_value.to_object(global_object);
-    VERIFY(this_object);
+    // 3. Let O be ! ToObject(this value).
+    auto* object = this_value.to_object(global_object);
+    VERIFY(object);
+
+    // 4. Let isArray be ? IsArray(O).
+    auto is_array = Value(object).is_array(global_object);
+    if (vm.exception())
+        return {};
+
+    String builtin_tag;
+
+    // 5. If isArray is true, let builtinTag be "Array".
+    if (is_array)
+        builtin_tag = "Array";
+    // 6. Else if O has a [[ParameterMap]] internal slot, let builtinTag be "Arguments".
+    else if (is<ArgumentsObject>(object))
+        builtin_tag = "Arguments";
```

Comment:
Hmm, this is not entirely correct, but im not sure how to fix it: this should print `Arguments` for any object with the `[[ParameterMap]]` internal slot, and the specification requries that we add an empty `[[ParameterMap]]` property to the unmapped arguments object as well (created by the CreateUnmappedArgumentsObject AO) so that it prints `Arguments` here as well

Reply:
I see, good catch... not sure either, and I don't want to mess with the argument objects right now, so I'll add a fixme :)

Reply:
scratch that, done!